benchmark/profile_restful_api.py

```diff
@@ -3,191 +3,233 @@
+    tokenized_dataset = []
+    for i in range(len(dataset)):
+        output_len = len(completion_token_ids[i])
+        tokenized_dataset.append((prompts[i], prompt_token_ids[i], output_len))
+
+    # Filter out too long sequences.
+    filtered_dataset: List[Tuple[str, int, int]] = []
+    for prompt, prompt_token_ids, output_len in tokenized_dataset:
+        prompt_len = len(prompt_token_ids)
+        if prompt_len < 4 or output_len < 4:
+            # Prune too short sequences.
+            continue
+        if prompt_len > 1024 or prompt_len + output_len > 2048:
+            # Prune too long sequences.
             continue
-        filtered_dataset.append([prompt, input_len, output_len])
+        filtered_dataset.append((prompt, prompt_len, output_len))
+
+    # Sample the requests.
+    sampled_requests = random.sample(filtered_dataset, num_requests)
+    return sampled_requests
+
 
-    if samples > 0:
-        filtered_dataset = random.sample(filtered_dataset, samples)
+class Engine:
 
-    que = Queue()
-    for data in filtered_dataset:
-        que.put(data)
-    que.put((None, None, None))
-    print(f'elapsed time for filtering: '
-          f'{round(time.perf_counter() - start, 2)} s')
-    return que, len(filtered_dataset)
+    def __init__(self,
+                 server_addr: str,
+                 tokenzier_path: str,
+                 temperature: float = 0.8,
+                 top_p: float = 1.0):
+        self.tokenizer = Tokenizer(tokenzier_path)
+        self.api_url = server_addr + '/v1/chat/completions'
+        self.temperature = temperature
+        self.top_p = top_p
+
+    def _inference(self, req_queue: Queue, res_queue: Queue, session_id: int,
+                   stream_output: bool):
+
+        stats = []
+        for prompt, input_seqlen, output_seqlen in iter(
+                req_queue.get, [None, None, None]):
+            timestamps = []
+            tokens = []
+            timestamps.append(time.perf_counter())
+            headers = {'content-type': 'application/json'}
+            pload = {
+                'model': 'llama',
+                'messages': prompt,
+                'temperature': self.temperature,
+                'top_p': self.top_p,
+                'n': 1,
+                'max_tokens': output_seqlen,
+                'stream': stream_output,
+                'session_id': session_id,
+                'ignore_eos': True,
+            }
+            response = requests.post(self.api_url,
+                                     headers=headers,
+                                     json=pload,
+                                     stream=stream_output)
+            for chunk in response.iter_lines(chunk_size=8192,
+                                             decode_unicode=False,
+                                             delimiter=b'\n'):
+                timestamps.append(time.perf_counter())
+                if chunk:
+                    data = json.loads(chunk.decode('utf-8'))
```

Comment:
may refer to https://github.com/InternLM/lmdeploy/blob/c07f60fd0abb76e6d6a066636b408630422a9f1f/lmdeploy/serve/openai/api_client.py#L96

Decoding `v1/chat/completions` is different from `generate`